src/main/java/io/supertokens/inmemorydb/Start.java

```diff
@@ -3032,9 +3032,27 @@ public void addClientForApp(AppIdentifier appIdentifier, String clientId)
 
             if (isPrimaryKeyError(serverErrorMessage, config.getOAuthClientTable(),
                     new String[]{"app_id", "client_id"})) {
-                throw new ClientAlreadyExistsForAppException();
+                throw new OAuth2ClientAlreadyExistsForAppException();
             }
             throw new StorageQueryException(e);
         }
     }
+
+    @Override
+    public boolean isClientIdAlreadyExists(String clientId) throws StorageQueryException {
+        try {
+            return OAuthQueries.isClientIdAlreadyExists(this, clientId);
+        } catch (SQLException e) {
+            throw new StorageQueryException(e);
+        }
+    }
```

Comment:
everything in the core is per app (if not per tenant) - so why do we need this function given that we already have doesClientIdExistForThisApp?

Reply:
This is used for checking if the client exists at all, not just for the app. @porcellus said that when registering a new client for an app that client is going to have a different clientID. So this is used when generating the clientId to check if that generated clientId is globally unique.

Reply:
Also, if we have one hydra instance, clientIds have to be unique within that hydra instance, so that means that we have to check for globall uniqueness here, not just for app

Reply:
Yea, but Hydra also does this check and returns a 409 anyway. So i see no reason that we also do this check

Reply:
I was thinking skipping the network roundtrip when we could detect this in place. But okay, if that's not a concern, I will remove this and the related parts.